gen/routing.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"os"
+	"path"
+
+	log "github.com/ipfs/go-log"
+	"github.com/ipld/edelweiss/compile"
+	"github.com/ipld/edelweiss/defs"
+)
+
+var proto = defs.Defs{
+
+	// delegated routing service definition
+	defs.Named{
+		Name: "DelegatedRouting",
+		Type: defs.Service{
+			Methods: defs.Methods{
+				defs.Method{
+					Name: "GetP2PProvide",
+					Type: defs.Fn{
+						Arg:    defs.Ref{Name: "GetP2PProvideRequest"},
+						Return: defs.Ref{Name: "GetP2PProvideResponse"},
+					},
+				},
+			},
+		},
+	},
+
+	// request type
+	defs.Named{
+		Name: "GetP2PProvideRequest",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{
+					Name: "Keys",
+					Type: defs.Named{
+						Name: "KeyList",
+						Type: defs.List{Element: defs.Ref{Name: "Multihash"}},
+					},
+				},
+			},
+		},
+	},
+
+	// response type
+	defs.Named{
+		Name: "GetP2PProvideResponse",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{
+					Name: "ProvidersByKey",
+					Type: defs.Named{
+						Name: "ProvidersByKeyList",
+						Type: defs.List{Element: defs.Ref{Name: "ProvidersByKey"}},
+					},
+				},
+			},
+		},
+	},
+	defs.Named{
+		Name: "ProvidersByKey",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{
+					Name: "Key",
+					Type: defs.Ref{Name: "Multihash"},
+				},
+				defs.Field{
+					Name: "Provider",
+					Type: defs.Ref{Name: "Provider"},
+				},
+			},
+		},
+	},
+
+	// general routing types
+	defs.Named{
+		Name: "Multihash",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{
+					Name: "Bytes",
+					Type: defs.Bytes{},
+				},
+			},
+		},
+	},
+
+	defs.Named{
+		Name: "Provider",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{
+					Name: "Nodes",
+					Type: defs.List{Element: defs.Ref{Name: "Node"}},
+				},
+				defs.Field{
+					Name: "Proto",
+					Type: defs.List{Element: defs.Ref{Name: "TransferProto"}},
+				},
+			},
+		},
+	},
+
+	defs.Named{
+		Name: "Node",
+		Type: defs.Inductive{
+			Cases: defs.Cases{
+				defs.Case{Name: "Peer", Type: defs.Ref{Name: "Peer"}},
+			},
+		},
+	},
+
+	defs.Named{
+		Name: "Peer",
+		Type: defs.Structure{
+			Fields: defs.Fields{
+				defs.Field{Name: "ID", Type: defs.Bytes{}},
+				defs.Field{Name: "Multiaddresses", Type: defs.List{Element: defs.Bytes{}}},
+			},
+		},
+	},
+
+	defs.Named{
+		Name: "TransferProto",
+		Type: defs.Inductive{
+			Cases: defs.Cases{
+				defs.Case{Name: "Bitswap", Type: defs.Ref{Name: "BitswapTransfer"}},
+			},
+		},
```

Comment:
Are there docs for this definition format? it seems fairly different from ipld schemas, which is what I was expecting to find here.
I think this is where I'm supposed to comment that there needs to be a graphsync case in addition to bitswap.

More generally, if providers push data to the indexer that is not 'bitswap' (or now '0x900'), how do we represent other transferprotos that have not been explicitly defined in a schema here?

Reply:
We just finished the RPC compiler, so indeed there is no doc. I am working on it now. "Inductive" is the new name for what used to be called "Union" in IPLD schema. ("Inductive" is the widely-accepted name for this type, "union" is something very different.) Otherwise, the type is the same: You have string-valued tags (the case name), which can carry values of any type.

The transfer protocol differentiator can be implemented in two different ways. Using the "inductive" type:
```go
defs.Inductive{
   Cases: defs.Cases{
      defs.Case{Name: "Bitswap", Type: defs.Ref{Name: "BitswapTransfer"}},
      defs.Case{Name: "Graphsync", Type: defs.Ref{Name: "GraphsyncTransfer"}},
   },
},
```

Or using the "Union" type and the "Singleton" type. The "union" type is a new type (that does not exist in IPLD schema). It differentiates cases based on the (deep) structure of the case subtypes (not based on tags, tag names are just for visual convenience when using unions, they are not reflected on the wire). "Singleton" means that only one specific value can inhabit this type.

```
defs.Union{
      defs.Case{Name: "Bitswap", Type: defs.Ref{Name: "BitswapTransfer"}},
      defs.Case{Name: "Graphsync", Type: defs.Ref{Name: "GraphsyncTransfer"}},
}

Named{
  Name: "BitswapTransfer",
  Type: Structure{
     Fields: Fields{
        Field{Name: "Magic", Type: SingletonInt{0x900}}
        ...
     }
  }
}

Named{
  Name: "GraphsyncTransfer",
  Type: Structure{
     Fields: Fields{
        Field{Name: "Magic", Type: SingletonInt{0xA00}}
        ...
     }
  }
}
```

In this example, the bitswap and the graphsync cases are differentiated because their "magic" fields are always different.

"Union" types are not implemented because they were not part of MVP, however if you end up needing them, they are a couple of days worth of work.